universe/ShipDesign.cpp

```diff
@@ -84,6 +84,52 @@ namespace {
         // have an ID yet) is the same as one that has already been added
         // (which does have an ID)
     }
+
+    constexpr unsigned int CHECKSUM_MODULUS = static_cast<unsigned int>(10'000'000);
+
+    template <class C>
+    void CheckSumCombine(unsigned int& sum, const C& c) {
+        for (const C::value_type& t : c)
+            CheckSumCombine(sum, t);
+
+        sum += c.size();
+        sum %= CHECKSUM_MODULUS;
+    }
+
+    template <>
+    void CheckSumCombine(unsigned int& sum, const bool& t) {
+        sum += t;
+        sum %= CHECKSUM_MODULUS;
+    }
+
+    template <>
+    void CheckSumCombine(unsigned int& sum, const char& t) {
+        sum += t;
+        sum %= CHECKSUM_MODULUS;
+    }
+
+    template <>
+    void CheckSumCombine(unsigned int& sum, const unsigned int& t) {
+        sum += t;
+        sum %= CHECKSUM_MODULUS;
+    }
+
+    template <>
+    void CheckSumCombine(unsigned int& sum, const int& t) {
+        sum += std::abs(t);
+        sum %= CHECKSUM_MODULUS;
+    }
+
+    template <>
+    void CheckSumCombine(unsigned int& sum, const double& t) {
+        // biggest and smallest possible float should be ~10^(+/-308)
+        // taking log gives a number in the range +/- 309
+        // adding 400 gives numbers in the range ~0 to 800
+        // multiplying by 10'000 gives numbers in the range ~0 to 8'000'000
+        sum += static_cast<unsigned int>((std::log(std::abs(t)) + 400.0) * 10'000);
```

Comment:
they might, but as above, hopefully it won't matter, since nothing in the FOCS should parse to such a high-valued float constant.